File: src/lessons/welcome/array/basics/Array1Entity.scala
package lessons.welcome.array.basics;

import java.awt.Color;
import jlm.core.model.Game

class ScalaArray1Entity extends jlm.universe.bugglequest.SimpleBuggle {
	override def setX(i: Int)  {
		if (isInited)
			throw new RuntimeException(Game.i18n.tr("I'm sorry Dave, I'm affraid I can't let you use setX() in this exercise."));
	}
	override def setY(i: Int)  { 
		if (isInited)
			throw new RuntimeException(Game.i18n.tr("I'm sorry Dave, I'm affraid I can't let you use setY() in this exercise."));
	}
	override def setPos(x: Int, y:Int)  { 
		if (isInited)
			throw new RuntimeException(Game.i18n.tr("I'm sorry Dave, I'm affraid I can't let you use setPos() in this exercise."));
	}

	/* BEGIN TEMPLATE */
	def run() {
		/* BEGIN SOLUTION */
		val colors = new Array[Color](getWorldHeight());

		/* read the colors */
		for (i <- 0 to getWorldHeight()-1) {
			colors(i) = getGroundColor()
					forward();
		}

		/* duplicate the pattern */
		for (col <- 1 to getWorldWidth()-1) {
			left();
			forward();
			right();
			forward();
			makeLine(colors);
		}
	}

	def makeLine(colors:Array[Color]) {
		for (i <- 0 to getWorldWidth()-1) {
			mark(colors(i));
			forward();
		}
	}
	
	def mark(c:Color){
		setBrushColor(c);
		brushDown();
		brushUp();
		/* END SOLUTION */
	}
	/* END TEMPLATE */
}

File: src/lessons/welcome/array/basics/Array1Entity.c
#include "RemoteBuggle.h"

void run();
void makeLine(Color colors[]);
void mark(Color c);

int main(){
	run();
	return 0;
}


#line 1 "Array1"
/* BEGIN TEMPLATE */
void run() {
	/* BEGIN SOLUTION */
	Color colors[] = malloc(sizeof(Color)*getWorldHeight());
	int i;
	/* read the colors */
	for (i=0;i<getWorldHeight();i++) {
		colors[i]=getGroundColor();
		forward(1);
	}

	/* duplicate the pattern */
	for (i=1; i<getWorldWidth();i++) {
		left();
		forward(1);
		right();
		forward(1);
		makeLine(colors);
	}
}
void makeLine(Color colors[]) {
	int i;
	for (i=0;i<getWorldWidth();i++) {
		mark(colors[i]);
		forward(1);
	}
}
void mark(Color c){
	setBrushColor(c);
	brushDown();
	brushUp();
	/* END SOLUTION */
}
/* END TEMPLATE */
